Assignment-18
NN

In [1]:
import numpy as np
import pandas as pd

In [2]:
df=pd.read_csv('/content/sonardataset.csv')

In [3]:
df.head()

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60,Y
0,0.0200,0.0371,0.0428,0.0207,0.0954,0.0986,0.1539,0.1601,0.3109,0.2111,...,0.0027,0.0065,0.0159,0.0072,0.0167,0.0180,0.0084,0.0090,0.0032,R
1,0.0453,0.0523,0.0843,0.0689,0.1183,0.2583,0.2156,0.3481,0.3337,0.2872,...,0.0084,0.0089,0.0048,0.0094,0.0191,0.0140,0.0049,0.0052,0.0044,R
2,0.0262,0.0582,0.1099,0.1083,0.0974,0.2280,0.2431,0.3771,0.5598,0.6194,...,0.0232,0.0166,0.0095,0.0180,0.0244,0.0316,0.0164,0.0095,0.0078,R
3,0.0100,0.0171,0.0623,0.0205,0.0205,0.0368,0.1098,0.1276,0.0598,0.1264,...,0.0121,0.0036,0.0150,0.0085,0.0073,0.0050,0.0044,0.0040,0.0117,R
4,0.0762,0.0666,0.0481,0.0394,0.0590,0.0649,0.1209,0.2467,0.3564,0.4459,...,0.0031,0.0054,0.0105,0.0110,0.0015,0.0072,0.0048,0.0107,0.0094,R


In [4]:
df.shape

(208, 61)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x_1     208 non-null    float64
 1   x_2     208 non-null    float64
 2   x_3     208 non-null    float64
 3   x_4     208 non-null    float64
 4   x_5     208 non-null    float64
 5   x_6     208 non-null    float64
 6   x_7     208 non-null    float64
 7   x_8     208 non-null    float64
 8   x_9     208 non-null    float64
 9   x_10    208 non-null    float64
 10  x_11    208 non-null    float64
 11  x_12    208 non-null    float64
 12  x_13    208 non-null    float64
 13  x_14    208 non-null    float64
 14  x_15    208 non-null    float64
 15  x_16    208 non-null    float64
 16  x_17    208 non-null    float64
 17  x_18    208 non-null    float64
 18  x_19    208 non-null    float64
 19  x_20    208 non-null    float64
 20  x_21    208 non-null    float64
 21  x_22    208 non-null    float64
 22  x_

In [6]:
df.describe()

,x_1,x_2,x_3,x_4,x_5,x_6,x_7,x_8,x_9,x_10,...,x_51,x_52,x_53,x_54,x_55,x_56,x_57,x_58,x_59,x_60
count,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,...,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000,208.000000
mean,0.029164,0.038437,0.043832,0.053892,0.075202,0.104570,0.121747,0.134799,0.178003,0.208259,...,0.016069,0.013420,0.010709,0.010941,0.009290,0.008222,0.007820,0.007949,0.007941,0.006507
std,0.022991,0.032960,0.038428,0.046528,0.055552,0.059105,0.061788,0.085152,0.118387,0.134416,...,0.012008,0.009634,0.007060,0.007301,0.007088,0.005736,0.005785,0.006470,0.006181,0.005031
min,0.001500,0.000600,0.001500,0.005800,0.006700,0.010200,0.003300,0.005500,0.007500,0.011300,...,0.000000,0.000800,0.000500,0.001000,0.000600,0.000400,0.000300,0.000300,0.000100,0.000600
25%,0.013350,0.016450,0.018950,0.024375,0.038050,0.067025,0.080900,0.080425,0.097025,0.111275,...,0.008425,0.007275,0.005075,0.005375,0.004150,0.004400,0.003700,0.003600,0.003675,0.003100
50%,0.022800,0.030800,0.034300,0.044050,0.062500,0.092150,0.106950,0.112100,0.152250,0.182400,...,0.013900,0.011400,0.009550,0.009300,0.007500,0.006850,0.005950,0.005800,0.006400,0.005300
75%,0.035550,0.047950,0.057950,0.064500,0.100275,0.134125,0.154000,0.169600,0.233425,0.268700,...,0.020825,0.016725,0.014900,0.014500,0.012100,0.010575,0.010425,0.010350,0.010325,0.008525
max,0.137100,0.233900,0.305900,0.426400,0.401000,0.382300,0.372900,0.459000,0.682800,0.710600,...,0.100400,0.070900,0.039000,0.035200,0.044700,0.039400,0.035500,0.044000,0.036400,0.043900


In [7]:
df.isnull().sum()

,0
x_1,0
x_2,0
x_3,0
x_4,0
x_5,0
...,...
x_57,0
x_58,0
x_59,0
x_60,0


In [8]:
df.duplicated().sum()

np.int64(0)

In [9]:
df.dtypes

,0
x_1,float64
x_2,float64
x_3,float64
x_4,float64
x_5,float64
...,...
x_57,float64
x_58,float64
x_59,float64
x_60,float64


In [10]:
df.columns

Index(['x_1', 'x_2', 'x_3', 'x_4', 'x_5', 'x_6', 'x_7', 'x_8', 'x_9', 'x_10',
       'x_11', 'x_12', 'x_13', 'x_14', 'x_15', 'x_16', 'x_17', 'x_18', 'x_19',
       'x_20', 'x_21', 'x_22', 'x_23', 'x_24', 'x_25', 'x_26', 'x_27', 'x_28',
       'x_29', 'x_30', 'x_31', 'x_32', 'x_33', 'x_34', 'x_35', 'x_36', 'x_37',
       'x_38', 'x_39', 'x_40', 'x_41', 'x_42', 'x_43', 'x_44', 'x_45', 'x_46',
       'x_47', 'x_48', 'x_49', 'x_50', 'x_51', 'x_52', 'x_53', 'x_54', 'x_55',
       'x_56', 'x_57', 'x_58', 'x_59', 'x_60', 'Y'],
      dtype='object')

In [11]:
X = df.drop('Y',axis=1)
y = df['Y']

In [12]:
X.shape,y.shape

((208, 60), (208,))

In [13]:
df.duplicated().sum()

np.int64(0)

In [14]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
y=le.fit_transform(y)

In [15]:
from sklearn.model_selection import train_test_split

In [16]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

In [17]:
X_train.shape,X_test.shape,y_train.shape,y_test.shape

((166, 60), (42, 60), (166,), (42,))

In [18]:
from sklearn.preprocessing import StandardScaler
sc=StandardScaler()
X_train=sc.fit_transform(X_train)
X_test=sc.transform(X_test)

In [19]:
X_train.shape,X_test.shape,y_train.shape,y_test.shape

((166, 60), (42, 60), (166,), (42,))

In [20]:
from tensorflow.keras.models import Sequential

from tensorflow.keras.layers import Dense

In [21]:
model= Sequential()

In [22]:
model.add(Dense(32, activation='relu', input_dim=60))
model.add(Dense(16, activation='relu'))
model.add(Dense(1, activation='sigmoid'))

/usr/local/lib/python3.13/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [23]:
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

In [24]:
model.fit(X_train,y_train, epochs=100, batch_size=8)

Epoch 1/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 4s 12ms/step - accuracy: 0.5422 - loss: 0.6927
Epoch 2/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - accuracy: 0.7831 - loss: 0.5908
Epoch 3/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.8253 - loss: 0.5266
Epoch 4/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.8735 - loss: 0.4660
Epoch 5/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.8855 - loss: 0.4136
Epoch 6/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.9096 - loss: 0.3641
Epoch 7/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.9277 - loss: 0.3202
Epoch 8/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 0.9337 - loss: 0.2823
Epoch 9/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 11ms/step - accuracy: 0.9518 - loss: 0.2483
Epoch 10/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 13ms/step - accuracy: 0.9699 - loss: 0.2216
Epoch 11/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.9699 - loss: 0.1969
Epoch 12/100
21/21 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - a

In [25]:
y_pred= model.predict(X_test)

2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 102ms/step


In [26]:
y_pred =(y_pred>0.5).astype(int)

In [27]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)

Accuracy: 0.8809523809523809


In [66]:
from sklearn.metrics import recall_score

recall = recall_score(y_test, y_pred)

print("Recall:", recall)

Recall: 0.875


In [63]:
from sklearn.metrics import precision_score

precision = precision_score(y_test, y_pred)

print("Precision:", precision)

Precision: 0.8235294117647058


In [28]:
from sklearn.metrics import confusion_matrix

Confusion_Matrix = confusion_matrix(y_test, y_pred)

print("Confusion Matrix:")
print(Confusion_Matrix)

Confusion Matrix:
[[23  3]
 [ 2 14]]


In [29]:
from sklearn.metrics import classification_report

Classification_rep = classification_report(y_test, y_pred)

print("Classification Report:")
print(Classification_rep)

Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.88      0.90        26
           1       0.82      0.88      0.85        16

    accuracy                           0.88        42
   macro avg       0.87      0.88      0.88        42
weighted avg       0.88      0.88      0.88        42



In [30]:
from sklearn.metrics import confusion_matrix, classification_report

cm = confusion_matrix(y_test, y_pred)

print("Confusion Matrix:")
print(cm)

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Confusion Matrix:
[[23  3]
 [ 2 14]]

Classification Report:
              precision    recall  f1-score   support

           0       0.92      0.88      0.90        26
           1       0.82      0.88      0.85        16

    accuracy                           0.88        42
   macro avg       0.87      0.88      0.88        42
weighted avg       0.88      0.88      0.88        42



In [31]:
from sklearn.model_selection import GridSearchCV

In [32]:
def create_model(neurons):
    model = Sequential()
    model.add(Dense(neurons, activation='relu', input_dim=60))
    model.add(Dense(16,activation='relu'))
    model.add(Dense(1, activation='sigmoid'))
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    return model

In [33]:
!pip install scikeras

In [34]:
from scikeras.wrappers import KerasClassifier

In [35]:
param_grid = {
    'model__neurons': [16, 32],
    'epochs': [50, 100],
    'batch_size': [8, 16]
}

In [36]:
param_grid

{'model__neurons': [16, 32], 'epochs': [50, 100], 'batch_size': [8, 16]}

In [37]:
model = KerasClassifier(
    model=create_model,
    verbose=0
)


In [38]:
grid = GridSearchCV(
    estimator=model,
    param_grid=param_grid,
    cv=3,
    scoring='accuracy'
)

In [39]:
import sklearn
import scikeras

print("scikit-learn:", sklearn.__version__)
print("scikeras:", scikeras.__version__)

scikit-learn: 1.5.2
scikeras: 0.13.0


In [40]:
!pip install scikit-learn==1.5.2

In [41]:
grid_result = grid.fit(X_train, y_train)

/usr/local/lib/python3.13/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.13/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr/local/lib/python3.13/dist-packages/keras/src/layers/core/dense.py:106: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/usr

In [43]:
print(type(X_train))
print(X_train.shape)

print(type(y_train))
print(y_train.shape)

<class 'numpy.ndarray'>
(166, 60)
<class 'numpy.ndarray'>
(166,)


In [45]:
print(create_model)

<function create_model at 0x7984d3d8ed40>


In [47]:
print("Best Parameters:", grid_result.best_params_)
print("Best CV Accuracy:", grid_result.best_score_)

Best Parameters: {'batch_size': 16, 'epochs': 100, 'model__neurons': 32}
Best CV Accuracy: 0.8075757575757576


In [48]:
print(grid_result.best_params_)

{'batch_size': 16, 'epochs': 100, 'model__neurons': 32}


In [51]:
best_model = grid_result.best_estimator_

y_pred_tuned = best_model.predict(X_test)

y_pred_tuned = (y_pred_tuned > 0.5).astype(int)


In [52]:
from sklearn.metrics import accuracy_score

accuracy_tuned = accuracy_score(y_test, y_pred_tuned)

print("Tuned Model Accuracy:", accuracy_tuned)

Tuned Model Accuracy: 0.8333333333333334


In [54]:
from sklearn.metrics import confusion_matrix, classification_report

cm_tuned = confusion_matrix(y_test, y_pred_tuned)

print("Tuned Model Confusion Matrix:")
print(cm_tuned)

Tuned Model Confusion Matrix:
[[20  6]
 [ 1 15]]


In [64]:
from sklearn.metrics import precision_score

precision_tuned = precision_score(y_test, y_pred_tuned)

print("Tuned Model Precision:", precision_tuned)

Tuned Model Precision: 0.7142857142857143


In [55]:
from sklearn.metrics import confusion_matrix, classification_report

cr_tuned = classification_report(y_test, y_pred_tuned)

print("Tuned Model Classification Report:")
print(cr_tuned)

Tuned Model Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.77      0.85        26
           1       0.71      0.94      0.81        16

    accuracy                           0.83        42
   macro avg       0.83      0.85      0.83        42
weighted avg       0.86      0.83      0.84        42



In [56]:
from sklearn.metrics import f1_score

In [57]:
f1_tunded = f1_score(y_test, y_pred_tuned)

print("Tuned Model F1 Score:", f1_tunded)
#

Tuned Model F1 Score: 0.8108108108108109


In [68]:
from sklearn.metrics import recall_score

recall_tuned = recall_score(y_test, y_pred_tuned)

print("Tuned Model Recall:", recall_tuned)


Tuned Model Recall: 0.9375


In [69]:
print("Tuned Model Accuracy:", accuracy_tuned)
print("Tuned Model F1 Score:", f1_tunded)
print("Tuned Model Confusion Matrix:")
print(cm_tuned)
print("Tuned Model Classification Report:")
print(cr_tuned)

Tuned Model Accuracy: 0.8333333333333334
Tuned Model F1 Score: 0.8108108108108109
Tuned Model Confusion Matrix:
[[20  6]
 [ 1 15]]
Tuned Model Classification Report:
              precision    recall  f1-score   support

           0       0.95      0.77      0.85        26
           1       0.71      0.94      0.81        16

    accuracy                           0.83        42
   macro avg       0.83      0.85      0.83        42
weighted avg       0.86      0.83      0.84        42



In [70]:
print("Baseline Accuracy:", accuracy)
print("Tuned Accuracy:", accuracy_tuned)

Baseline Accuracy: 0.8809523809523809
Tuned Accuracy: 0.8333333333333334


In [71]:
tuned_accuracy = accuracy_score(y_test, y_pred_tuned)
tuned_precision = precision_score(y_test, y_pred_tuned)
tuned_recall = recall_score(y_test, y_pred_tuned)
tuned_f1 = f1_score(y_test, y_pred_tuned)
print("Tuned Accuracy:", tuned_accuracy)
print("Tuned Precision:", tuned_precision)
print("Tuned Recall:", tuned_recall)
print("Tuned F1 Score:", tuned_f1)

Tuned Accuracy: 0.8333333333333334
Tuned Precision: 0.7142857142857143
Tuned Recall: 0.9375
Tuned F1 Score: 0.8108108108108109


In [74]:
print("Baseline Accuracy:", baseline_accuracy)
print("Baseline Precision:", baseline_precision)
print("Baseline Recall:", baseline_recall)
print("Baseline F1:", baseline_f1)
print("Tuned Accuracy:", tuned_accuracy)
print("Tuned Precision:", tuned_precision)
print("Tuned Recall:", tuned_recall)
print("Tuned F1:", tuned_f1)

Baseline Accuracy: 0.8809523809523809
Baseline Precision: 0.8235294117647058
Baseline Recall: 0.875
Baseline F1: 0.8484848484848485
Tuned Accuracy: 0.8333333333333334
Tuned Precision: 0.7142857142857143
Tuned Recall: 0.9375
Tuned F1: 0.8108108108108109


In [75]:
print("Best Parameters:", grid_result.best_params_)
print("Best CV Accuracy:", grid_result.best_score_)

Best Parameters: {'batch_size': 16, 'epochs': 100, 'model__neurons': 32}
Best CV Accuracy: 0.8075757575757576


The baseline ANN achieved 88.10% accuracy, while the tuned ANN achieved 83.33%. The tuned model had higher recall (9), but the baseline model performed better overall in accuracy, precision, and F1-score.

The ANN successfully classified sonar signals as mine or rock. Hyperparameter tuning showed that different settings affect model performance. In this experiment, the baseline model performed better overall.